src/slowlog.c

```diff
@@ -0,0 +1,160 @@
+            return CORVUS_ERR;
+        }
+    }
+    return CORVUS_OK;
+}
+
+// Should be called only after all worker threads have stopped
+void slowlog_free(struct slowlog_queue *slowlog)
+{
+    for (size_t i = 0; i != slowlog->capacity; i++) {
+        // no other worker threads will read entry queue any more
+        if (slowlog->entries[i] != NULL) {
+            slowlog_dec_ref(slowlog->entries[i]);
+        }
+        pthread_mutex_destroy(slowlog->entry_locks + i);
+    }
+    cv_free(slowlog->entries);
+    cv_free(slowlog->entry_locks);
+}
+
+struct slowlog_entry *slowlog_create_entry(struct command *cmd, int64_t latency)
+{
+    struct slowlog_entry *entry = cv_calloc(1, sizeof(struct slowlog_entry));
+    entry->id = ATOMIC_INC(slowlog_id, 1);
+    entry->log_time = time(NULL);  // redis also uses time(NULL);
+    entry->latency = latency;
+    entry->refcount = 1;
+
+    const char *bytes_fmt = "(%zd bytes)";
+    char bytes_buf[19];  // strlen("4294967295") + strlen("( bytes)") + 1
+
+    assert(cmd->data.elements > 0);
+    size_t argc = min(cmd->data.elements, SLOWLOG_ENTRY_MAX_ARGC);
+    entry->argc = argc;
+    size_t i = 0;
+
+    // for sub cmd
+    if (cmd->prefix) {
+        const char *c = cmd_extract_prefix(cmd->prefix);
+        assert(c != NULL);
+        entry->argv[i].len = strlen(c);
+        entry->argv[i++].str = (uint8_t*)c;
+        entry->argc = min(argc + 1, SLOWLOG_ENTRY_MAX_ARGC);
+        argc = min(argc, SLOWLOG_ENTRY_MAX_ARGC - 1);
+    }
+    // use the last argument to record total argument counts
+    if (cmd->data.elements > argc) {
+        argc--;
+        const size_t tail_max_len = 64;
+        char tail_buf[tail_max_len];
+        uint8_t *buf = cv_malloc(tail_max_len);
+        int len = snprintf(tail_buf, tail_max_len, "(%zd arguments in total)", cmd->data.elements);
+        len = snprintf((char*)buf, tail_max_len, "$%d\r\n%s\r\n", len, tail_buf);
+        entry->argv[SLOWLOG_ENTRY_MAX_ARGC - 1].str = buf;
+        entry->argv[SLOWLOG_ENTRY_MAX_ARGC - 1].len = len;
+    }
+    size_t max_len = SLOWLOG_ENTRY_MAX_STRING;
+    for (size_t j = 0; j != argc; i++, j++) {
+        struct redis_data *arg = cmd->data.element + j;
+        assert(arg->type == REP_STRING);
+        size_t real_len = mbuf_range_len(arg->buf);
+        size_t len = min(real_len, max_len);
+        uint8_t *buf = cv_malloc(len);  // will be freed by the last owner
+        if (real_len > max_len) {
+            size_t str_len = SLOWLOG_MAX_ARG_LEN;
+            int hdr_len = snprintf((char*)buf, len, "$%zd\r\n", str_len);
+            pos_to_str_with_limit(&arg->pos, buf + hdr_len, str_len);
+            int postfix_len = snprintf(bytes_buf, sizeof bytes_buf, bytes_fmt, arg->pos.str_len);
+            memcpy(buf + max_len - 2 - postfix_len, bytes_buf, postfix_len);
+            memcpy(buf + max_len - 2, "\r\n", 2);
+        } else {
+            mbuf_range_copy(buf, arg->buf, len);
```

Comment:
Because I want copy the whole bulk string here, including the `$len\r\n` prefix and a trailing `\r\n`